tools/mutation_test/mutator.py

```diff
@@ -0,0 +1,499 @@
+        " signed " : " unsigned ",
+        " unsigned " : " signed ",
+        " long " : 
+            [ " int ", " short int ", " char " ],
+        " float ": " int ",
+        " double ": " int ",
+
+
+        " free(": "// free(",
+
+        "case ": "// case ",
+        "default ": "// default ",
+
+        # null terminate a string
+        "\"": "\"\\0",
+
+        "else {": "{",
+        "else": "// else",
+    }
+}
+
+
+class Occurrence():
+    """ Object to keep track of a pattern's occurrence.
+
+    Access the object's fields directly.
+
+    `pattern`: a Pattern type.
+    `file`: str containing path to the file (source code).
+    `line`: int containing the line in the file where the `pattern` is found.
+    `index`: int representing the index of the line where the `pattern` occurs.
+    """
+    def __init__(self, pattern, file, line, index):
+        self.pattern = pattern
+        self.file = file
+        self.line = line
+        self.index = index
+    
+    def __str__(self):
+        return f'Pattern {self.pattern.pattern} found at File: {self.file}, Line: {self.line}, Index: {self.index}'
+    
+    def __repr__(self):
+        return f'({self.pattern}, {self.file}, {self.line}, {self.index})'
+
+class Pattern():
+    """ Represents a mutation pattern
+    
+    Access fields directly.
+
+    `pattern`: str containing the original mutant operator.
+
+    `transformation`: str containing the mutated operator.
+
+    eg. `pattern`: " = ", `transformation`: " <= "
+    """
+    def __init__(self, pattern, transformation):
+        self.pattern = pattern
+        self.transformation = transformation
+    
+    def __str__(self):
+        return f'{self.pattern} : {self.transformation}'
+
+class Mutator():
+    """ The main driver object to create mutations for a run. 
+    
+    Create one Mutator object for each set of patterns and source code files.
+    """
+
+    def __init__(self, src: dict, mutation_patterns: dict, rng=random.Random()):
+        """ Initializes various fields
+        
+        `src` is a dictionary mapping from a path to file to a list of line ranges. eg.
+
+        { 
+            "code.c" : [],
+            "code2.c" : [[1,12],[50,150]]
+        }
+
+        `mutation_pattern` is a dictionary mapping from str to list of str or str. eg.
+
+        {
+            " == " : [ " <= ", " != " ],
+            " return " : [" return -2 * ", " return -1 + "],
+            "++" : "--"
+        }
+
+        `rng` is a Random type. Used for random operations.
+        """
+        def _merge(intervals):
+            """ Turns `intervals` [[0,2],[1,5],[7,8]] to [[0,5],[7,8]].
+            """
+            out = []
+            for i in sorted(intervals, key=lambda i: i[0]):
+                if out and i[0] <= out[-1][1]:
+                    out[-1][1] = max(out[-1][1], i[1])
+                else:
+                    out += i,
+            return out
+        
+        def _flatten(intervals):
```

Comment:
move this function to utils ?

Reply:
https://github.com/aws/amazon-freertos/pull/2475/commits/76f5a7464e42986f2d3702cbdf4fcb180e50da3f